README.rst
.. image:: https://img.shields.io/badge/python-3.7-blue

.. image:: https://github.com/dzhw/mdm-python-client/workflows/Build/badge.svg
    :target: https://github.com/dzhw/mdm-python-client/actions

.. image:: https://codecov.io/gh/dzhw/mdm-python-client/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/dzhw/mdm-python-client

.. image:: https://readthedocs.org/projects/mdm-python-client/badge/?version=latest
    :target: https://mdm-python-client.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

=================
MDM Python Client
=================

This project contains a REST API client for `MDM <https://metadata.fdz.dzhw.eu>`_  (`Metadatamanagement System <https://github.com/dzhw/metadatamanagement>`_) of the Research Data Center of the DZHW.

Description
===========

The client can be used to retrieve metadata from the public `REST API <https://metadata.fdz.dzhw.eu/api/swagger-ui.html>`_ of the MDM_. Currently it uses a cache which gets stale after 24 hours.

You can install the latest version of this client with

.. code:: bash

    pip install git+https://github.com/dzhw/mdm-python-client.git#v1.0.106

The previous last version which supports Python 2.7 can be installed with

.. code:: bash

    pip install git+https://github.com/dzhw/mdm-python-client.git#v1.0.104

The installation will also install a command line version which you can run with

.. code:: bash

    mdm --help

The following snippet demonstrates how to use the client programmatically:

.. code:: python

    from mdm_python_client.mdm_client import get_datapackage

    def test_get_datapackage():
        datapackage = get_datapackage("stu-gra2005$")
        if datapackage is not None:
            print("Pakettitel:", datapackage.title.de)
            print("DOI:", datapackage.doi)
            print("Veröffentlichungsjahr:", datetime.datetime.strptime(datapackage.release.firstDate,
                                                               '%Y-%m-%dT%H:%M:%S.%f').strftime("%Y"))
        else:
            print("Datenpaket mit id", args.datapackageId, "nicht gefunden!")

Having trouble?
===============

Please file an issue in the `issue tracker <https://github.com/dzhw/metadatamanagement/issues>`_ of the `MDM`_.


CHANGES.rst
=========
Changelog
=========

Version 0.0.1
=============

- Quick implementation of an MDM client including caching

Version 0.0.2
=============

- Move cache to temp directory

Version 0.0.3
=============

- Try to detect directory with proper access rights for our CMS

Version 1.0.88
==============

First production release:

- Client uses persistent cache which is valid for 24 hours.
- It falls back to in-memory cache if the file for storing the cache cannot be written.
- In case of any (network) error the cached result (if present) is returned.

Version 1.0.104
===============

- Change study to data package
- Fix issues with python 3.7
- upgrade dependencies

Version 1.0.106
===============

- Migrate to Github Actions
- Drop Python 2.7 support


tests/conftest.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for mdm_python_client.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""
from __future__ import print_function, absolute_import, division

import pytest


tests/test_main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from mdm_python_client.main import main
from mdm_python_client.mdm_client import get_datapackage, get_instrument

__author__ = "René Reitmann"
__copyright__ = "René Reitmann"
__license__ = "AGPL-3.0"

test_api_endpoint = 'https://test.metadata.fdz.dzhw.eu/api'


def test_main_with_data_package_id():
    main(["--datapackage", "stu-gra2005$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_data_package_id():
    main(["--datapackage", "hurz", "--endpoint", test_api_endpoint])


def test_main_with_instrument_id():
    main(["--instrument", "ins-gra2005-ins1$", "--endpoint", test_api_endpoint])


def test_main_with_survey_id():
    main(["--survey", "sur-gra2005-sy1$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_survey_id():
    main(["--survey", "hurz", "--endpoint", test_api_endpoint])


def test_main_with_invalid_instrument_id():
    main(["--instrument", "hurz", "--endpoint", test_api_endpoint])
    
    
def test_main_with_question_id():
    main(["--question", "que-gra2005-ins1-1.1$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_question_id():
    main(["--question", "hurz", "--endpoint", test_api_endpoint])


def test_main_with_dataset_id():
    main(["--dataset", "dat-gra2005-ds1$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_dataset_id():
    main(["--dataset", "hurz", "--endpoint", test_api_endpoint])
    

def test_main_with_variable_id():
    main(["--variable", "var-gra2005-ds1-astu011a$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_variable_id():
    main(["--variable", "hurz", "--endpoint", test_api_endpoint])


def test_main_with_publication_id():
    main(["--publication", "pub-Kratz.2016$", "--endpoint", test_api_endpoint])


def test_main_with_invalid_publication_id():
    main(["--publication", "hurz", "--endpoint", test_api_endpoint])


tests/test_skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import pytest
from mdm_python_client.skeleton import fib

__author__ = "René Reitmann"
__copyright__ = "René Reitmann"
__license__ = "none"


def test_fib():
    assert fib(1) == 1
    assert fib(2) == 1
    assert fib(7) == 13
    with pytest.raises(AssertionError):
        fib(-10)


mdm_python_client/skeleton.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
     fibonacci = mdm_python_client.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""
from __future__ import division, print_function, absolute_import

import argparse
import sys
import logging

from mdm_python_client import __version__

__author__ = "René Reitmann"
__copyright__ = "René Reitmann"
__license__ = "none"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonnaci demonstration")
    parser.add_argument(
        '--version',
        action='version',
        version='mdm-python-client {ver}'.format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


mdm_python_client/mdm_client.py
# coding=utf-8
import json
import logging
import os
import tempfile
from collections import namedtuple

import requests_cache


# use a sqllite DB for caching the MDM calls for 24 hours
def __setup_cache():
    # check several directories for write access for the sqllite db
    directories = {tempfile.gettempdir(), "/usr/local/plone_local_files", os.environ.get("TMPDIR", None),
                   os.environ.get("TEMPDIR", None)}
    for cache_dir in directories:
        if cache_dir is not None and os.path.exists(cache_dir) and os.access(cache_dir, os.W_OK):
            print("Going to write cached requests into file:", os.path.join(cache_dir, "mdm_client_cache"
                                                                                       ".sqlite"))
            return requests_cache.CachedSession(cache_name=os.path.join(cache_dir, "mdm_client_cache"),
                                                backend="sqlite", expire_after=60 * 60 * 24, old_data_on_error=True)
    # use memory as fallback
    print("Going to use in-memory cache!")
    return requests_cache.CachedSession(cache_name="mdm_client_cache",
                                        backend="memory", expire_after=60 * 60 * 24, old_data_on_error=True)


_session = __setup_cache()


def __json_object_hook(d): return namedtuple("X", d.keys())(*d.values())


def __json2obj(data): return json.loads(data, object_hook=__json_object_hook)


def __get_object(url):
    """
    Get the json from the given URL and convert it into a python object.

    :param url: The complete url for the GET request.
    :return: A python object created from the response, or None if not found.
    """
    response = _session.get(url)
    # For successful API call, response code will be 200 (OK)
    if response.ok:
        # Loads (Load String) takes a Json file and converts into python data structure (dict)
        return __json2obj(response.content)
    elif response.status_code == 404:
        return None
    else:
        # If response code is not ok (200), print the resulting http error code with description
        response.raise_for_status()


# baseUrl = "https://metadatamanagement-test.cfapps.io/api"
baseUrl = "https://metadata.fdz.dzhw.eu/api"


def get_datapackage(datapackage_id, endpoint=baseUrl):
    """
    Get the data package with the given datapackage_id from the MDM. Results will be cached for 24 hours in order to speed up
    subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param datapackage_id: The id of the data package (e.g. "stu-gra2005$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/data-packages/" + datapackage_id
    return __get_object(url)


def get_survey(survey_id, endpoint=baseUrl):
    """
    Get the survey with the given survey_id from the MDM. Results will be cached for 24 hours in order to speed up
    subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param survey_id: The id of the survey (e.g. "sur-gra2005-sur1$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/surveys/" + survey_id
    return __get_object(url)


def get_instrument(instrument_id, endpoint=baseUrl):
    """
    Get the instrument with the given instrument_id from the MDM. Results will be cached for 24 hours in order to speed
    up subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param instrument_id: The id of the instrument (e.g. "ins-gra2005-ins1$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/instruments/" + instrument_id
    return __get_object(url)


def get_question(question_id, endpoint=baseUrl):
    """
    Get the question with the given question_id from the MDM. Results will be cached for 24 hours in order to speed up
    subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param question_id: The id of the question (e.g. "que-gra2005-ins1-1.1$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/questions/" + question_id
    return __get_object(url)


def get_dataset(dataset_id, endpoint=baseUrl):
    """
    Get the dataset with the given dataset_id from the MDM. Results will be cached for 24 hours in order to speed
    up subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param dataset_id: The id of the dataset (e.g. "dat-gra2005-ds1$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/data-sets/" + dataset_id
    return __get_object(url)


def get_variable(variable_id, endpoint=baseUrl):
    """
    Get the variable with the given variable_id from the MDM. Results will be cached for 24 hours in order to speed
    up subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param variable_id: The id of the variable (e.g. "var-gra2005-ds1-astu011a$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/variables/" + variable_id
    return __get_object(url)


def get_concept(concept_id, endpoint=baseUrl):
    """
    Get the concept with the given concept_id from the MDM. Results will be cached for 24 hours in order to speed
    up subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param concept_id: The id of the concept (e.g. "con-personality$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/concepts/" + concept_id
    return __get_object(url)


def get_publication(publication_id, endpoint=baseUrl):
    """
    Get the publication with the given publication_id from the MDM. Results will be cached for 24 hours in order to
    speed up subsequent calls. A Rest API explorer can be found here: https://metadata.fdz.dzhw.eu/api/swagger-ui.html

    :param publication_id: The id of the publication (e.g. "pub-personality$")
    :param endpoint: The url of the MDM Api endpoint (default: "https://metadata.fdz.dzhw.eu/api")
    :return: A python object created from the response, or None if not found.
    """
    url = endpoint + "/related-publications/" + publication_id
    return __get_object(url)


mdm_python_client/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following line in the
entry_points section in setup.cfg:

    console_scripts =
     mdm = mdm_python_client.main:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.
"""
from __future__ import division, print_function, absolute_import

import argparse
import datetime
import logging
import sys

from mdm_python_client import __version__, mdm_client

__author__ = "René Reitmann"
__copyright__ = "DZHW GmbH"
__license__ = "AGPL-3.0"

_logger = logging.getLogger(__name__)


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a MDM Rest API Client")
    parser.add_argument(
        '--version',
        action='version',
        version='mdm-python-client {ver}'.format(ver=__version__))
    parser.add_argument(
        '--datapackage',
        dest="datapackageId",
        help="The id of a data package of the MDM (e.g. stu-gra2005$)",
        type=str,
        metavar="datapackageId")
    parser.add_argument(
        '--instrument',
        dest="instrumentId",
        help="The id of an instrument of the MDM (e.g. ins-gra2005-ins1$)",
        type=str,
        metavar="instrumentId")
    parser.add_argument(
        '--question',
        dest="questionId",
        help="The id of an question of the MDM (e.g. que-gra2005-ins1-1.1$)",
        type=str,
        metavar="questionId")
    parser.add_argument(
        '--survey',
        dest="surveyId",
        help="The id of an survey of the MDM (e.g. sur-gra2005-sy1$)",
        type=str,
        metavar="surveyId")
    parser.add_argument(
        '--dataset',
        dest="datasetId",
        help="The id of a dataset of the MDM (e.g. dat-gra2005-ds1$)",
        type=str,
        metavar="dataSetId")
    parser.add_argument(
        '--variable',
        dest="variableId",
        help="The id of a variable of the MDM (e.g. var-gra2005-ds1-astu011a$)",
        type=str,
        metavar="variableId")
    parser.add_argument(
        '--publication',
        dest="publicationId",
        help="The id of a publication of the MDM (e.g. pub-Kratz.2016$)",
        type=str,
        metavar="publicationId")
    parser.add_argument(
        '--concept',
        dest="conceptId",
        help="The id of a concept of the MDM (e.g. con-personality$)",
        type=str,
        metavar="conceptId")
    parser.add_argument(
        '--endpoint',
        dest="endpoint",
        help="The API endpoint of the MDM (e.g. https://metadata.fdz.dzhw.eu/api)",
        type=str,
        metavar="url",
        default="https://metadata.fdz.dzhw.eu/api")
    parser.add_argument(
        '-v',
        '--verbose',
        dest="loglevel",
        help="set loglevel to INFO",
        action='store_const',
        const=logging.INFO)
    parser.add_argument(
        '-vv',
        '--very-verbose',
        dest="loglevel",
        help="set loglevel to DEBUG",
        action='store_const',
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting MDM Rest API Client...")
    if args.datapackageId is not None:
        _logger.debug("Fetching data package %s", args.datapackageId)
        datapackage = mdm_client.get_datapackage(args.datapackageId, args.endpoint)
        if datapackage is not None:
            print("Pakettitel:", datapackage.title.de)
            print("DOI:", datapackage.doi)
            print("Veröffentlichungsjahr:", datetime.datetime.strptime(datapackage.release.firstDate,
                                                                       '%Y-%m-%dT%H:%M:%S.%f').strftime("%Y"))
        else:
            print("Datenpaket mit id", args.datapackageId, "nicht gefunden!")
    if args.surveyId is not None:
        _logger.debug("Fetching survey %s", args.surveyId)
        survey = mdm_client.get_survey(args.surveyId, args.endpoint)
        if survey is not None:
            print("Erhebungstitel:", survey.title.de)
        else:
            print("Erhebung mit id ", args.surveyId, "nicht gefunden!")
    if args.instrumentId is not None:
        _logger.debug("Fetching instrument %s", args.instrumentId)
        instrument = mdm_client.get_instrument(args.instrumentId, args.endpoint)
        if instrument is not None:
            print("Instrumenttitel:", instrument.title.de)
        else:
            print("Instrument mit id ", args.instrumentId, "nicht gefunden!")
    if args.questionId is not None:
        _logger.debug("Fetching question %s", args.questionId)
        question = mdm_client.get_question(args.questionId, args.endpoint)
        if question is not None:
            print("Fragetext:", question.questionText.de)
        else:
            print("Frage mit id ", args.questionId, "nicht gefunden!")
    if args.datasetId is not None:
        _logger.debug("Fetching dataset %s", args.datasetId)
        dataset = mdm_client.get_dataset(args.datasetId, args.endpoint)
        if dataset is not None:
            print("Datensatzbeschreibung:", dataset.description.de)
        else:
            print("Datensatz mit id ", args.datasetId, "nicht gefunden!")
    if args.variableId is not None:
        _logger.debug("Fetching variable %s", args.variableId)
        variable = mdm_client.get_variable(args.variableId, args.endpoint)
        if variable is not None:
            print("Variablenlabel:", variable.label.de)
        else:
            print("Variable mit id ", args.variableId, "nicht gefunden!")
    if args.publicationId is not None:
        _logger.debug("Fetching publication %s", args.publicationId)
        publication = mdm_client.get_publication(args.publicationId, args.endpoint)
        if publication is not None:
            print("Publikationstitel:", publication.title)
        else:
            print("Publikation mit id ", args.publicationId, "nicht gefunden!")
    if args.conceptId is not None:
        _logger.debug("Fetching concept %s", args.conceptId)
        concept = mdm_client.get_concept(args.conceptId, args.endpoint)
        if concept is not None:
            print("Konzepttitel:", concept.title.de)
        else:
            print("Konzept mit id ", args.conceptId, "nicht gefunden!")
    _logger.info("MDM REST API Client finished")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()
